core/l1_fee_context.go

```diff
@@ -0,0 +1,100 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"bytes"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/predeploys"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var big10 = big.NewInt(10)
+
+var (
+	L1BaseFeeSlot = common.BigToHash(big.NewInt(1))
+	OverheadSlot  = common.BigToHash(big.NewInt(3))
+	ScalarSlot    = common.BigToHash(big.NewInt(4))
+	DecimalsSlot  = common.BigToHash(big.NewInt(5))
+)
+
+// calculateL1GasUsed returns the gas used to include the transaction data in
+// the calldata on L1
+func calculateL1GasUsed(data []byte, overhead *big.Int) *big.Int {
+	var zeroes uint64
+	var ones uint64
+	for _, byt := range data {
+		if byt == 0 {
+			zeroes++
+		} else {
+			ones++
+		}
+	}
+
+	zeroesGas := zeroes * params.TxDataZeroGas
+	onesGas := (ones + 68) * params.TxDataNonZeroGasEIP2028
+	l1Gas := new(big.Int).SetUint64(zeroesGas + onesGas)
+	return new(big.Int).Add(l1Gas, overhead)
+}
+
+// L1FeeContext includes all the context necessary to calculate the cost of
+// including the transaction in L1
+type L1FeeContext struct {
+	BaseFee  *big.Int
+	Overhead *big.Int
+	Scalar   *big.Int
+	Decimals *big.Int
+	Divisor  *big.Int
+}
+
+// NewL1FeeContext returns a context for calculating L1 fee cst
+func NewL1FeeContext(cfg *params.ChainConfig, statedb *state.StateDB) *L1FeeContext {
+	// TODO: unpack values after #2596
+	// see: https://github.com/ethereum-optimism/optimism/pull/2596
+	l1BaseFee := statedb.GetState(common.HexToAddress(predeploys.L1Block), L1BaseFeeSlot).Big()
+	overhead := statedb.GetState(common.HexToAddress(predeploys.OVM_GasPriceOracle), OverheadSlot).Big()
+	scalar := statedb.GetState(common.HexToAddress(predeploys.OVM_GasPriceOracle), ScalarSlot).Big()
+	decimals := statedb.GetState(common.HexToAddress(predeploys.OVM_GasPriceOracle), DecimalsSlot).Big()
+	divisor := new(big.Int).Exp(big10, decimals, nil)
+
+	return &L1FeeContext{
+		BaseFee:  l1BaseFee,
+		Overhead: overhead,
+		Scalar:   scalar,
+		Decimals: decimals,
+		Divisor:  divisor,
+	}
+}
+
+// L1Cost returns the L1 fee cost.
+// This depends on the chainconfig because gas costs
+// can change over time
+func L1Cost(tx *types.Transaction, ctx *L1FeeContext) *big.Int {
+	var rlp bytes.Buffer
+	if err := tx.EncodeRLP(&rlp); err != nil {
```

Comment:
Not super opinionated but have seen people moving towards `tx.MarshalBinary`  since typed txs got introduced, https://github.com/ethereum/go-ethereum/blob/67454df08b512ea33aa4c246ee405cf4b418ec35/core/types/transaction.go#L114

Reply:
`MarshalBinary` has fewer bytes than `EncodeRLP`. I think we want to use the exact same encoding scheme we'll be using in the op-proposer / op-batcher?

Reply:
iirc they are the same, they messed with the RLP method to handle the non-RLP type byte

Reply:
this should be switched over to `Encode` or `MarshalBinary`. If how the batches are encoded in the rollup node add the extra bytes, it should be fixed there or we should update the overhead.

Reply:
Switched to `MarshalBinary`.